doc/developer/design/20230806_durable_catalog_state.md

```diff
@@ -0,0 +1,402 @@
+        1. Open a persist read handle.
+        2. Read a snapshot of the catalog state into memory.
+        3. Run migrations in memory.
+            - None of the objects will be cached in memory.
+- `open_read_only`:
+    - If the catalog is not initialized then fail.
+    - If the catalog is initialized:
+        1. Open a persist read handle.
+        2. Read a snapshot of the catalog state into memory.
+- `boot_ts`: Return boot timestamp from memory.
+- `epoch`: This will be stored the same as any other catalog object so this can be treated as a
+  normal [read](#reads).
+
+#### Reads
+
+All reads (except for the deploy generation) will perform the follow steps:
+
+1. Confirm leadership.
+2. Return an object from the in memory cache.
+
+Reading the deploy generation will perform the following steps:
+
+- If the catalog is open, then read perform normal read.
+- If the catalog is not open:
+    - If the catalog is not initialized then return `None`.
+    - If the catalog is initialized:
+        1. Open a persist read handle.
+        2. Read the deploy generation from persist.
+
+Reading the deployment generation needs to work before the catalog has been opened and before
+migrations have been run.
+
+#### Writes
+
+- `transaction`: Use objects from in memory cache to initialize a new `Transaction`.
+- `commit_transaction`:
+    1. Use the write handle to compare and append the batch of changes into persist.
+    2. Update the in memory upper to the new upper.
+    3. Update in-memory object state.
+- `set_X`:
+    1. For each object create the following batch of changes:
+        - If the key of the object already exists in memory, add a retraction to the batch.
+        - Add an addition to the batch of the new value.
+    2. Use the write handle to compare and append the batch of changes into persist.
+    3. Update the in memory upper to the new upper.
+    4. Update in-memory object state.
+
+#### Misc
+
+- `confirm_leadership`: Check that the persist shard upper is equal to the upper cached in memory.
+- `dump`: Convert the in-memory state to a JSON string.
+
+### Catalog Transactions
+
+The current
+[catalog transactions](https://github.com/MaterializeInc/materialize/blob/b5aa928960aa68c031caa8ff3bfab0ff53220361/src/adapter/src/catalog/storage.rs#L892-L916)
+will remain mostly the same. It works by reading the entire durable catalog state into memory,
+making changes to the state in memory, then committing all changes at once to the backing durable
+storage. The few changes needed are:
+
+- It will be the responsibility of the `DurableCatalogState` to commit the `Transaction` instead of
+  the `Transaction` committing itself.
+- `Transaction` will not hold onto a mutable reference to a `Stash`. Though it will need a mutable
+  reference to some part of `DurableCatalogState` to prevent multiple concurrent transactions.
+- To avoid cloning the entire catalog state in the persist implementation, we'll
```

Comment:
I wonder if it's time to switch to immutable data structures that could handle this more elegantly than clone or cow. Certainly doesn't need to be now (cow is fine).